smart-contracts/contracts/BridgeBank/BridgeBank.sol

```diff
@@ -434,13 +434,57 @@ contract BridgeBank is BankStorage, CosmosBank, EthereumWhiteList, CosmosWhiteLi
       contractDecimals[token] = _decimals;
     } catch {
       // if we can't access the decimals function of this token,
-      // assume that it has 18 decimals
-      decimals = 18;
+      // assume that it has 0 decimals
+      decimals = 0;
     }
 
     return decimals;
   }
 
+  /**
+   * @dev Fetches the current token balance the bridgebank holds of a given token address
+   * @param token The bridgeToken's address
+   * @return The balance of the bridgebanks account with the bridge token
+   */
+  function getBalance(address token) private returns (uint256) {
+    uint256 balance;
+    try BridgeToken(token).balanceOf(address(this)) returns (uint256 _balance) {
+      balance = _balance;
+    } catch {
+      balance = 0;
+    }
+    return balance;
+  }
+
+  /**
+   * @dev Function which transfers the requested amount of tokens from the calling users account to the bridgebanks account
+   *      This function checks the balances before and after the transfer and reports the amount that was transfered in total
+   *      such that tokens which charge fees on transfer are accurately represented.
+   * @param token The bridgeToken's address
+   * @param amount The amount of bridgeToken's to transfer to the bridgebank
+   * @return The balance that was transfered as reported by the getBalance command   
+   */
+  function transferBalance(address token, uint256 amount) private returns (uint256) {
+    //The interface of the ERC20 token to interact with
+    IERC20 tokenToTransfer = IERC20(token);
+
+    //The balance before any transfers take place
+    uint256 oldBalance = getBalance(token);
+
+    // locking the tokens by transfering them from the user to the bridgebank
+    tokenToTransfer.safeTransferFrom(msg.sender, address(this), amount);
+
+    //Fetch the updated balance reported after the transfer
+    uint256 newBalance = getBalance(token);
+
+    //Calculate the total amount transfered from the newbalance vs the old balance
+    //Since this contract uses solidity 0.8+ overflows from bad acting tokens should 
+    //revert.
+    uint256 transferedAmount = newBalance - oldBalance;
+    
```

Comment:
Add the require > 0 here instead of in the callers?

Reply:
I was thinking about that as well, I wanted to have a more generic transfer function that does not make assumptions about what has to happen after the transfer occurs and so I coded it this way. That being said if there is never a time in our code we would not want that check then it makes more sense to include it in the transfer function. Do you think that's the best?

Reply:
Could there be any issues with concurrency here?

Reply:
@jzvikart As I understand it, there is no concurrency on standard EVM implementations such that it should not be a concern. If I am wrong please let me know, this is why I wanted as many eyes on the code as possible.

Reply:
@smartyalgo Unlock operations should be prevented on the ethbridge layer for values of 0 since those are costs incurred by the relayer. We should never revert on a valid call to unlock as that could cause problems with the ordering. As for the unlock event, an unlock request was received for and transferred out of n units, I don't know if we would want to show an event for less than what the ethbridge module created a claim for.